daemon/cosign.conf.5
.TH cosign.conf "5" "March 2012" "cosign" "File Formats Manual"
.SH NAME
.B cosign.conf
\- Cookie Sign-On Daemons and CGI configuration file
.SH SYNOPSIS
.B _COSIGN_CONF
.sp
.SH DESCRIPTION
The _COSIGN_CONF file is a general purpose configuration file for
cosign.cgi, the logout cgi, cosignd, and monster. It is a whitespace delimited
line-oriented file. The first argument is always a keyword, followed by
a value, followed by optional per-keyword additional values. The
keywords
.B cgi
and
.B proxy
are unique to cosignd. The keywords
.BR service ,
.B set
and
.B include
are used by both daemons and both cgis. Finally the keywords
.BR cert ,
.BR factor ,
.BR negotiate ,
.B reauth
and
.B suffix
are used by cosign.cgi.
.sp
Keywords and values are whitespace-delimited. Lines that are blank or begin
with '#' are ignored (see Examples).
.sp
.SH CGI and SERVICE
The
.B cgi
and
.B service
keywords describe the role of the client within the cosign system.
.sp
The
.B cgi
keyword indicates that the client is responsible for running the cosign.cgi.
.B cgi
takes one argument, a regular expression of the subject CN (common name)
of the certificate the client will present to cosignd when connecting.
.sp
The
.B service
keyword indicates that the client is running a cosign-protected web service.
.B service
takes 4 or 5 arguments, depending on the effect the administrator wishes to
achieve. cosignd examines all of the fields; cosign.cgi looks only at the
first three.
.sp
The first argument is a regular expression (regex) of the service
name associated with this particular web service, for example,
"cosign-(web\\.mail)".  Argument 2 is the validation URL associated with
the service, which the service uses to validate the service cookie and
destination URL. This is the location to which cosign.cgi will redirect
following a successful authentication. The regex and the validation URL
are connected, in that matches in the regex can be substituted in the
validation URL (see Examples).
.sp
The third field contains flags (T, 2, 0) denoting whether the client is
permitted to retrieve Kerberos tickets (T); whether the client is permitted
to use the deprecated cosign v2 scheme (2); or whether the client will use
defaults (0). Flags can be combined.
.sp
The fourth field is a regular expression of the client's CN, on
which is based the client's access to cosignd,
or more specifically, to any cookies matching the service regex
given as the first argument. That is to say, only client CNs matching
the regex in the fourth argument are allowed to talk to cosignd 
about service cookies matching the regex in the first argument.
.sp
Generally the administrator will limit access to cookies by relying on
the association established between the service regex (argument 1) and
the client CN regex (argument 4), but there are cases where this
association is too vague. The optional fifth argument, a service name
substitution pattern, is provided for these cases. The canonical example
is a so-called "default" service, allowing institution-wide access to
cosignd without requiring the weblogin administrators to modify
.BR _COSIGN_CONF ,
as shown in this example:
.sp
.br
 service cosign-(.*) https://$1/cosign/valid 0 (.*\\.xyz\\.org) cosign-$1
.sp
This service entry allows any client with a valid CN ending in "xyz.org"
access to the server, but because of the presence of the substitution
pattern, those clients can only talk to cosignd
about cookies derived from their CN. For example, a client presenting a
certificate CN of "infiniteloop.xyz.org" would only be able to talk
to cosignd about "cosign-infiniteloop.xyz.org" service cookies.
.sp
Whether using the
.B cgi
keyword or the
.B service
keyword, a client's CN must match a line in the configuration file or it
will not be permitted to talk to cosignd. All clients must call STARTTLS
before any other commands can be issued. Clients can only issue commands
associated with their keyword. The keyword
.B cgi
allows clients to use the
.BR LOGIN ,
.BR CHECK ,
.BR REGISTER ,
.BR RETRIEVE ,
.BR TIME ,
.BR DAEMON ,
and
.B LOGOUT
commands. 
.sp
The keyword
.B service
allows clients to use the
.B CHECK
and
.B RETRIEVE
commands (see cosignd(8)).
.sp
.SH PROXY
The
.B proxy
keyword allows a given service to retrieve proxy cookies for other services.
.B proxy
takes two arguments, a client CN regex and a path to a proxy file. The
proxy file contains the hostname and servicename for the proxy cookies,
one per line, whitespace-delimited (see Examples). The CN regex must
match a prior service entry in
.BR _COSIGN_CONF .
The
.B proxy
keyword replaces the combination of the "P" flag with a path in older
versions of Cosign.
.sp
.SH "SET" CONFIG FILE OPTIONS - SHARED
The following configuration file options are recognised for use with the
keyword
.BR SET .
Configuration file options override the default compile time settings, but are in turn overridden by any command line options as specified.
.TP 19
.B cosigncadir
The location where CA certificates in hashed form after expected to reside for validating client ceriticates. This is overridden by the
.B \-x
command line option, and the default is _COSIGN_TLS_CADIR
.TP 19
.B cosigncert
The location of the SSL server certificate for use by this program. This is overridden by the
.B \-y
command line option and the defualt is _COSIGN_TLS_CERT
.TP 19
.BI cosignkey
The location of the SSL key file for use by this program. This is overridden by the
.B \-z
command line option and the default is _COSIGN_TLS_KEY
.TP 19
.BI cosigndcrl
The location of the SSL certificate revocation list (CRL). This path may be a CRL file or a hashed directory of CRLs. CRLs are not enabled by default. Setting
.B cosigndcrl 
enables them.
.SH "SET" CONFIG FILE OPTIONS - DAEMON
.TP 19
.B cosigndb
The cosign daemon directory, where all cookies (cosign= and cosign-serv=) will be stored. This is overridden by the
.B \-D
command line option and the default is _COSIGN_DIR
.TP 19
.B cosigndbhashlen
This is the length of the subdirectory hash in cosignd's cookie cache
_COSIGN_DIR. The default is 0, which means all cookies are written to
the top level of _COSIGN_DIR. A value of 1 means cosignd will use 64
subdirectories, that is a-z, A-Z, 0-9, +, and - (this is only 38 on a
case-insensitive filesystem) to store the cookie cache. A value of 2
means cosignd will use 4096 subdirectories, similar to thr previous
example but with 2 character directory names (this is only 1444 on a
case-insensitive filesystem) to store the cookie cache. These
subdirectories must be created by the administrator (see
SCRIPTS/dbhash).
.TP 19
.B cosignhost
The hostname to replicate to. This "turns on" cosignd's replication.
This is overridden by the
.B \-h
command line option and the default is off and "NULL"
.TP 19
.B cosigndticketcache
The path to the directory where cosignd stores the kerberos tickets sent
by cosign.cgi. If nothing is set here, the default value is _COSIGN_TICKET_CACHE
.TP 19
.B cosignnettimeout
This is the amount of time alotted by cosignd for each network
transaction. The default is 4 minutes.
.TP 19
.B cosignport
This is the port on which cosignd listens. The default is 6663.
.TP 19
.B cosignstrictcheck
This can be set to "on" or "off". Enable or disable strict limitations on
cookie access. If enabled, cosignd will only allow a client to check
cookies associated with the client's certificate Common Name. If
disabled, cosignd does not enforce any limitations on cookie checks.
The default value is "on", enabling strict checking.
.sp
.SH "SET" CONFIG FILE OPTIONS - CGIs
.TP 19
.B cosignkrbtkts
This can be set to "on" or "off". Governs whether an external 
HTTP-Negotiate or x509 login should attempt to transfer kerberos 
tickets. The default is "off".
.TP 19
.B cosignticketcache
The path to the directory where cosign.cgi stores the kerberos tickets
it generates upon successful kerberos login. These tickets are deleted
once the ticket has been successfully transfered to cosignd. This
setting is also the default path for the directory where cosignd stores
its tickets, and the default is _COSIGN_TICKET_CACHE
.TP 19
.B cosignhttponlycookies
Enable or disable the "httponly" attribute of the cosign cookie set by the 
.BR cosign.cgi .
The "httponly" attribute restricts non-User Agent access to the cookie,
mitigating some cross-site scripting (XSS) attacks. The value can be set
to "on" or "off". The default is "on".
.TP 19
.B cosigntmpldir
The path to the directory where cosign.cgi gets the templates for
drawing the screen. This is therefore the working directory for
cosign.cgi. The default is _COSIGN_TMPL_DIR
.TP 19
.B cosignx509krbtkts
Synonymous with cosignkrbtkts.
.TP 19
.B cosignkeytab
This is the path to the keytab file keytab with the principal of "cosign" and
the instance of the hostname of the machine that the cgi will run on. This
keytab is used during authentication to verify that KDC spoofing is not
happening  <http://www.securityfocus.com/bid/1616> and must therefore be
readable by the UID the CGI runs as (usually the same as your web server).
The path to this file can be set at compile time using the
.B --with-keytabpath=FILE
configure option, during runtime with this cosignkeytab setting or
.IR krb5_kt_default_name (3)
will be used at run time to identify the default keytab file for the uid
running the CGI. The default is "NULL".
.TP 19
.B cosignprincipal
The principal name to use with
.B cosignkeytab
option. If set,
.B cosign.cgi
will attempt to find this principal within the keytab set by
.B cosignkeytab
and use it to verify the user's credentials. The default principal name
.B cosign.cgi
uses is "cosign/<hostname_of_cgi_server>".
.TP 19
.B cosignstoretickets
Used by the cgi to determine whether kerberos tickets should be stored in
a ccache after a successful kerberos login. Valid values are
.B on
and
.BR off .
The default is
.BR on .
.TP 19
.B cosignticketlifetime
Sets the lifetime in seconds of the kerberos ticket-granting ticket (TGT)
created by a successful kerberos login. The default is 10 hours.
.TP 19
.B cosignpamservice
Sets the pam.conf service name used by
.B cosign.cgi
when using PAM to authenticate users. The default service name is
_COSIGN_PAM_SERVICE.
.TP 19
.B cosignlogouturl
This is the URL the logout cgi will redirect a user to upon successful
completion of logout. It can be overriden by passing a URL on the query
string to the logout cgi, as in
http://cosign.edu/cosign-bin/logout?http://go.here.now.edu
and the default is _COSIGN_LOGOUT_URL.
.TP 19
.B cosignlogoutregex
The regular expression matching URLs passed in the query string that the
.B logout
cgi will consider safe for redirection after a logout. If the URL included
on the query string does not match the pattern, _COSIGN_LOGOUT_URL will be
used. The default pattern is _COSIGN_LOGOUT_RE.
.TP 19
.B cosignloopurl
This is the page a browser is redirected to when cosign.cgi detects
there is a loop. The complied-in defaults define a loop as visiting
cosign.cgi more than 10 (MAXLOOPCOUNT) times in 30 (LOOPWINDOW) seconds.
The default is _COSIGN_LOOP_URL
.sp
.SH "SET" CONFIG FILE OPTIONS - CGI with Cosign Friend
The following three options are for use with Cosign Friend which is now a
separate release, though the cgi in this release of Cosign must be built
with friend support on.
.TP 19
.B mysqldb
The hostname of friend's database server. The default is _FRIEND_MYSQL_DB
.TP 19
.B mysqldbname
The name of friend database. The default is _FRIEND_MYSQL_DB_NAME
.TP 19
.B mysqluser
The login name to access the friend database. The default is
_FRIEND_MYSQL_LOGIN
.TP 19
.B mysqlpasswd
The password to access the friend database. The default is
_FRIEND_MYSQL_PASSWD
.sp
.SH INCLUDE
The keyword
.B include
allows an administrator to build a cosign configuration file by
aggregating the contents of several smaller files. The first argument is
the keyword followed by the path to the file to include.
.sp
.SH REAUTH
The keyword 
.B reauth
denotes a cosign service that requires reauthentication before a
.B REGISTER
can succeed.
.B reauth
requires at least one argument, a service cookie prefix like
"cosign-web\\.mail".
.sp
Beginning with Cosign 2.0, there is the option to indicate that multiple
factors are required for reauthentication. After the service cookie prefix
there can be up to 5 factor names. When a user is asked to re-authenticate,
s/he must supply the correct information to every factor that is
configured here in the single POST. If the user types any information
wrong, s/he must type in everything again.
.sp
If there are no factors configured for reauthentication, the cgi will
accept any factor the user can satisfy. Any
.B reauth
line requires a prior matching
.B service
line in the configuration file.
.B reauth
replaces the "cookie" keyword in older versions of Cosign.
.sp
.SH CERT
The keyword
.B cert
is used to authenticate a user with an x509 certificate. It takes 5
fields - keyword, issuer_dn, subject_dn, login, and factor (see
Examples).  The issuer_dn must be an exact match  of the environment
variable SSL_CLIENT_I_DN.  The next field, subject_dn, is a regular
expression which matches SSL_CLIENT_S_DN. The login and factor fields
may contain $1, $2, and/or $3, which are replaced with the corresponding
parenthesized subexpressions of the regular expression subject_dn.
.sp
.SH FACTOR
The keyword
.B factor
is used to invoke the external authenticator API. It takes the keyword,
the pathname to the executable, and a list of form-fields, maximum of 5
(see Examples).  If all listed form-fields contain posted data, then the
CGI invokes the external authenticator. The value of each form-field is
written to the external authenticator on stdin (file descriptor 0), one
per line, in the order they are listed in the configuration. If
authentication is successful, the external authenticator writes the
factor name on stdout (file descriptor 1) and exits with a value of 0.
If an error occurs, the external authenticator writes an error message
on stdout and exits with a value of 1. If the user's password has
expired, the external authenticator writes an error message on stdout
and exits with a value of 2. All other exit values are reserved for
future use.
.sp
There are currently two additional configuration options.  The -2 option
to the factor keyword means that this factor is only checked after
another (non -2) factor has been satisfied.  It is intended for use with
factors that are vulnerable to denial-of-service attacks due to repeated
authentication failures. The other configuration option uses the keyword
.I suffix.
.sp
.SH NEGOTIATE
The keyword
.B negotiate
is used to authenticate a user using HTTP-Negotiate authentication. It 
takes 4 fields - keyword, regexp, login, and factor.  The regular 
expression regexp is matched against the REMOTE_USER variable 
provided by the web server. The login and factor fields may contain $1,
$2, and/or $3, which are replaced with the corresponding parenthesized
subexpressions of the regular expression.
.sp
.SH PASSWD
The keyword
.B passwd
is used to control password based authentication of a user using the 
Kerberos, MySQL and PAM internal authenticators. Where this keyword is not
specified, usernames containing an '@' are authenticated through mysql,
all other usernames are authenticated with Kerberos. The keyword takes
5 fields -
.IR keyword ,
.IR type ,
.IR regexp ,
.I login
and
.IR factor .
The regular
expression regexp is matched against the username supplied by the user.
The login and factor fields may contain $1, $2 and/or $3 which are replaced
with the corresponding parethesized subexpressions of the regular expression.
The CGI uses the first password line whose regular expression matches the
username, and calls the internal authenticator specified by the 'type'
field (which may be 'kerberos', 'mysql' or 'pam'), with the login name
given by the login field. Should this login be successful, the user is
marked as being authenticated by the factor contained in the factor field.
You may have all three configuration types enabled simultaneously.
.sp
The default authentication type rules match the following
.B passwd
configuration lines:
.sp
.nf
    # Anything containing an @ is a friend request
    passwd mysql    (.+@.+)  $1 friend
    # Anything else gets handed to Kerberos
    passwd kerberos ([^@]+) $1 MY.REALM
.fi
.sp
The following allows user to specify accounts in the form user@realm, and 
allows both A.EXAMPLE.COM and B.EXAMPLE.COM to be used as supported realms.
.sp
.nf
    passwd kerberos (.*@A\\.EXAMPLE\\.COM) $1 A.EXAMPLE.COM
    passwd kerberos (.*@B\\.EXAMPLE\\.COM) $1 B.EXAMPLE.COM
.fi
.sp
To enable PAM processing for user authentication, add the following line
in _COSIGN_CONF:
.sp
.nf
	passwd pam	(.+)	$1	PAM
.fi
.sp
This tells the cgi to authenticate all users with PAM. Authenticated users
with this configuration will have the authentication factor "PAM". This may
be changed by replacing the last field with a different value.
.sp
By default, the cosign.cgi will use the PAM service "_COSIGN_PAM_SERVICE"
to authenticate.  You may override this with the
.B cosignpamservice
keyword.
.sp
.SH SUFFIX
The keyword
.I suffix
takes 2 arguments - the keyword and a string, such as "-junk" (see
Examples). This configuration
causes cosign.cgi to remove the "-junk" suffix from any
server-provided factors before comparison with required 
factors.  For example, if the cgi is checking the factor "OTP", and
ignores the suffix "-junk", and the user has authenticated 
with the factor "OTP-junk", then the cgi's authentication factor
requirements would be fulfilled. This option is paired with the cosign
filter's CosignFactorSuffix and CosignFactorSuffixIgnore and is
essentially here to support Michigan's phased roll out of one time
passwords.
.sp
.SH EXAMPLES

CHANGELOG
3.2.0rc2
	build: Add "make test" to test build integrity.
		Currently tests cosignd, cosign.cgi and logout cgi.
	build: Add "make server" to build only weblogin server components.
	build: Add "make install-server"
	build: Add DESTDIR support to Makefile.
		Patch from Sean Sweda <ssweda at barracuda dot com>.
	common: incorporate fixes from cosign 3.1.4.
	cgi: Check environment for alternate cosign.conf.
		Added for make test.
	
3.2.0rc1
	cgi: Add support for httponly cookies.
	cgi: Add validuser() to allow limited non-alphanumeric usernames.
	filters: Add support for httponly cookies.
	filters: Permit CosignAllowPublicAccess in .htaccess files.
		Patch from pgp at psu dot edu.
	filters: Use vhost servername as default value for CosignService.
		[Feature Request #2748312]
	filters: Prevent site configuration from caching cosign redirects.
	filters: CosignAllowValidationRedirect works with CosignNoAppendRedirect
	filters: Reorder calls to perror to avoid errno reset.
		Patch from Martin Sucha <sucha14 at st dot fmph dot uniba.sk>
	filters: NULL-terminate ap_pstrcat and apr_pstrcat.

3.1.4
	common: fix NULL dereference causing cosignd crash on config read.

3.1.3
	common: implicitly anchor regular expression matches.
		patch for COSIGN-VULN-2011-001. See:
                http://weblogin.org/cosign-vuln-2011-001.txt

3.1.2
        apache2 filter: detect apache 2.2, and instruct httpd to order
                mod_cosign ahead of mod_authz_host. older versions of
                the apache 2.2 filter incorrectly referred to mod_access.
        filters: fix hard-coded protocol version in read_scookie.
                Bug report from liamr at umich dot edu.

3.1.1
	daemon: correct format for arguments when handling CHECKs from
		clients using protocol 0.

3.1.0
	all: Cookie rekeying in apache and lighttpd filters.
	daemon: add banner capability list to work around hardcoded
		protocol version.
	filters: add capability list parsing.
	filters: add REKEY command. Replaces "CHECK <cookie> rekey" in
		older RCs.
	cgi: catch NULL return value from snet_getline_multi when storing
		krb tickets.
	cgi: add cosignprincipal config option. Based on a patch from matt
		at linuxbox dot com.
	cgi: add cosignstoretickets and cosignticketlifetime config options.
	cgi: add cosignlogoutregex config option.
	common: fix: --enable-mysql's optional path argument was ignored.
		Based on a patch from jorj at isc dot upenn dot edu.
	common: Adopt autoheader conventions in build.
	common: Handle NetBSD's krb5 pathing in configure script. Report
		from nmadura at umich dot edu.
	filters: [Patch 2801877]: Support multi-cert PEMs. Patch fromfedora
		dot dm0 at gmail dot com.
	filters: IP checking now defaults to "never".
	filters: [Request 2748342]: Allow CosignService value to start with
		"cosign-".
	filters: fix length passed to strncasecmp when checking cookie
		prefix in filters.
	filters: fix regression in return value check for ap_pregcomp.
		Report from jorj at isc dot upenn dot edu.
	lighttpd filter: fix size checks when copying into sinfo struct.
	build: Make LIBS environment available to apxs for compiling.
	Additional build modifications to clean up git detritus

3.0.2
	filter: use ap_pregcomp to compile regexes

3.0.1
	daemon: add cosignstrictcheck configuration option
	filter: register handler correctly for apache1
	filter: allow extended regular expressions in CosignValidReference

3.0.0
	common: implement cosign v3 scheme to address COSIGN-VULN-2009-002,
		see http://weblogin.org/cosign-vuln-2009-002.txt

2.1.1
	cgi: escape login name to address COSIGN-VULN-2009-001,
		see http://weblogin.org/cosign-vuln-2009-001.txt

2.1.0
	common: improved configure option passing
	common: added --enable-universal-binaries
	cgi: escape factors (xss)
	cgi: use SERVER_NAME env variable rather than cosign_host config
	cgi: support SPNEGO (Simon Wilkinson)
	cgi: try passwords against multiple legacy methods (Simon Wilkinson)
	cgi: default authenticator regexes use extended format
        cgi: report expired passwords to user (Kris Steinhoff)
	cgi: report killed factors to user
	cgi: moved Friend password incorrect message handling to main loop
	cgi: don't escape ';'
	cgi: bug in certificate-based authN variable substitution (Petr Lampa)
	daemon: bug in kerberos ticket path handling (Petr Lampa)
	daemon: bug in non-replication restart code (Petr Lampa)
	daemon: improvement to replication error case under high load
	filter: improved apxs 2 usage
	filter: close connection more reliably (Petr Lampa)
	filter: detect race where multple HTTP GETs cause overlapping
		cosign CHECKS
	filter: fix connection reordering bug (Simon Wilkinson)
        filter: fix setting COSIGN_FACTORS from cached cookie
        filter: store server protocol version in cached cookie
	filter: return from cosign_cookie_valid() if cosign_check_cookie()
		returns anything other than COSIGN_OK (Petr Lampa)
	filter: improve IP checking

2.0.2a
	cgi: examine cookies for invalid character (security)
	filter: move IP checking

2.0.2
	cgi & filters: improved validation of login & service cookies
	cgi: better error messages
	cgi: improved building
	daemon: removed config logging, corrects openlog/syslog ordering problem
	monster: corrected bug in 2-character hashing code
	filter: improved SSL error handling
	filter: allow CosignRequireFactor in .htaccess
	filter: fixed bug that would "lose" Kerberos tickets in some cases
	filter: fixed bug causing failure to update cookie cache
	filter: fixed bug that incorrectly enables IP checking
	filter: switched cookie cache reading code from stdio to snet

2.0.1
	cgi: preauth failed and expired password are now re-tryable errors
	daemon: krbtktpath is now maintained upon subsequent logins
	filters: fixed some typos
	
2.0.0
	updates to READMES and docs
	common: general bug fixes, more helpful logging and error messages
	common: support for multifactor
	common: support for factor suffix 

html/js/cosign.js
// Determines if a value exists in an array
// based on code from embimedia.com
Array.prototype.inArray = function ( value )
{
    for ( var i=0; i<this.length; i++ ) {
		if ( typeof( value ) == 'object' ) {
			for ( var j=0; j<value.length; j++ ) {
				if ( this[i] === value[j] ) {
					return true;
				}
			}
		} else {
			// Matches identical (===), not just similar (==).
			if ( this[i] === value ) {
				return true;
			}
		}
    }

    return false;
};


//Remember object based on moo.fx by Valerio Proietti 
//and code by Peter-Paul Koch
var Remember = new Object();
Remember = function(){};

Remember.prototype = {
	setCookie: function( name, value ) {
		document.cookie = name + '=' + value + '; expires=Sun, 17 Jan 2038 12:34:56 UTC; path=/';
	},

	readCookie: function( name ) {
		var nameEQ = name + "=";
		var ca = document.cookie.split( ';' );
		for( var i=0; i<ca.length; i++ ) {
			var c = ca[i];
			while ( c.charAt( 0 )==' ' ) c = c.substring( 1, c.length );
			if ( c.indexOf( nameEQ ) == 0 ) {
				return {data: c.substring( nameEQ.length, c.length ), isset: true};
			}
		}

		return {data: '', isset: false};
	}
}

var Cosign = Class.create();
Cosign.prototype = Object.extend( new Remember(), {
	initialize: function( requiredFctrs ) {
		this.factorCookie    = 'exposedFactors';
		this.attribsCookie   = 'miscAttributes';
		this.factors         = Array();
		this.infoBoxes       = Array();
		this.requiredFctrs   = Array();
		this.cookieFactors   = Array();
		this.cookieAttribs   = Array();
		this.satisfied       = Array();
		this.defaultOpen     = Array();
		this.requiredFctrs   = requiredFctrs.split( ',' );
		var cookieFactors    = this.readCookie( this.factorCookie );
		this.cookieFactors   = cookieFactors.data.split( ',' );
		this.factorCookieSet = cookieFactors.isset;
		var cookieAttribs    = this.readCookie( this.attribsCookie );
		this.cookieAttribs   = cookieAttribs.data.split( ',' );
	},

	setSatisfied: function( satisfied ) {
		this.satisfied = satisfied.split( ',' );
	},

	setDefaultOpen: function( defaults ) {
		this.defaultOpen = defaults.split( ',' );
	},

	addFactor: function ( factorName, title, factorDivID ) {
		var focusBox = $( factorDivID ).getElementsByTagName( 'input' )[0];
		if ( focusBox.value != '' && focusBox.type != 'checkbox' ) {
			focusBox = $( factorDivID ).getElementsByTagName( 'input' )[1];
		}

		this.factors[title] = {factorName: factorName, title: title, factorDivID: factorDivID, focusBox: focusBox};
	},

	addInfoBox: function ( infoID, infoShow, infoHide ) {
		this.infoBoxes[infoShow] = {infoID: infoID, infoShow: infoShow, infoHide: infoHide};
	},
	
	setSubmitLink: function ( elName, formName ) {
		var submitLink = $( elName );
		Event.observe( submitLink, 'click', function() {$( formName ).submit()}, false );

		// This is old school, but Event.stop will not work in Safari as of 2.0.3...
		submitLink.onclick = function() {return false};
	},

	hide: function( factor ) {
		Effect.BlindUp( factor.factorDivID, {duration: 0.25} );
		$( factor.title ).className = 'dink';
	},

	showSatisfied: function( factor ) {
		Effect.BlindUp( factor.factorDivID, {duration: 0.25} );
		$( factor.title ).className = 'check';
		var nodes  = $( factor.title ).parentNode.childNodes;

		$A( nodes ).each(
			function( node ) {
				if ( node.className == 'authComplete' ) {
					Element.show( node );
				}
			}
		);
	},

	giveFocus: function( focusBox ) {
		try {
			$( focusBox ).focus();
			return true;
		} catch ( e ) {
			return false;
		}
	},

	toggle: function( factor ) {
		var blindStart = function() {
			if ( $( factor.factorDivID ).style.display == 'none' ) {
				$( factor.title ).className = 'dinkDown';
			} else {
				$( factor.title ).className = 'dink';
			}
		};

		var blindFinish = function() {
			this.saveFactorState( factor );
			this.closeInfo();
		}.bind(this);

		Effect.toggle( factor.factorDivID, 'blind', {duration: 0.25, beforeStart: blindStart, afterFinish: blindFinish} );
	},

	// Based on http://www.howtocreate.co.uk/jslibs/capsDetect.js
	capsDetect: function ( e ) {
		if( !e ) {
			e = window.event;
		}

		if( !e ) {
			this.MWJ_say_Caps( false );
			return;
		}

		//what (case sensitive in good browsers) key was pressed
		var theKey = e.which ? e.which : ( e.keyCode ? e.keyCode : ( e.charCode ? e.charCode : 0 ));
		//if upper case, check if shift is not pressed. if lower case, check if shift is pressed
		this.MWJ_say_Caps(( theKey > 64 && theKey < 91 && ! e.shiftKey ) || ( theKey > 96 && theKey < 123 && e.shiftKey ));
	},

	MWJ_say_Caps: function( oC ) {
		var elements = document.getElementsByClassName( 'capsLock' );
		if( oC ) {
			elements.each( function( node ) {Element.show( node )} );
		} else {
			elements.each( function( node ) {Element.hide( node )} );
		}
	},

	initUI: function() {
		var oThis    = this;
		var focusSet = false;
		
		// Don't set focus if user has already clicked or typed
		Event.observe( window, 'click', function() {focusSet = true} );
		Event.observe( window, 'keydown', function() {focusSet = true} );

		// escape framesets
		if ( window != top ) {
			top.location.href = location.href;
		}

		if ( $( 'error' ) != null ) {
			new Effect.Highlight( 'error' );
		}
		
		var nodeList  = document.getElementsByTagName( 'input' );
		var nodes     = $A( nodeList );

		nodes.each( function( node ) {
				// The Scriptaculous Event.Observe method cannot be used here.
				// Most browsers will not capture the event properly with it.
				if ( node.type == 'password') {
					node.onkeypress = function( e ){oThis.capsDetect( e )};
					node.setAttribute( 'autocomplete', 'off' );
				} else if ( node.type == 'text' ) {
					node.setAttribute( 'autocomplete', 'off' );
				}
			});

		// Set factor visibility and focus
		$H( this.factors ).each( function( factor ) {
			if ( this.satisfied.inArray( factor.value.factorName )) {
				this.showSatisfied( factor.value );
			} else if ( this.requiredFctrs.inArray( factor.value.factorName )) {
				$( factor.value.title ).className = 'required';
				if ( ! focusSet ) {
					focusSet = this.giveFocus( factor.value.focusBox );
				}
			} else if ( this.cookieFactors.inArray( factor.value.factorName )) {
				Event.observe(factor.key, 'click', function(){oThis.toggle(oThis.factors[factor.key])});
				if ( ! focusSet ) {
					focusSet = this.giveFocus( factor.value.focusBox );
				}
			} else if ( this.defaultOpen.inArray( factor.value.factorName ) && this.factorCookieSet == false ) {
				Event.observe(factor.key, 'click', function(){oThis.toggle(oThis.factors[factor.key])});
				if ( ! focusSet ) {
					focusSet = this.giveFocus( factor.value.focusBox );
				}

				// Set a cookie for the default open factor
				this.saveFactorState( factor.value );
			} else {
				Event.observe(factor.key, 'click', function(){oThis.toggle(oThis.factors[factor.key])});
				this.hide( factor.value );
			}
		}.bind( this ));

		// Register InfoBox click events
		$H( this.infoBoxes ).each( function( infoBox ) {
			Event.observe(infoBox.key, 'click', function(){oThis.showInfo(oThis.infoBoxes[infoBox.key].infoID)});
			Event.observe(infoBox.value.infoHide, 'click', function(){oThis.closeInfo()});
			Element.hide( infoBox.value.infoID );
			$( infoBox.value.infoID ).style.visibility = 'visible'; // Prevents a "flash" of all infoboxes on load
		});

		// Apply misc element attributes
		for ( var i=0; i<this.cookieAttribs.length; i++ ) {
			var attrib = this.cookieAttribs[i].split( '~' );
			if ( attrib.length == 3 ) {
				var el = $( attrib[0] );
				if ( el != null ) {
					el.setAttribute( attrib[1], attrib[2] );
				}
			}
		}
	},

	saveFactorState: function( factor ) {
		if ( $( factor.factorDivID ).style.display != 'none' && ! this.cookieFactors.inArray( factor.factorName )) {
			if ( typeof( factor.factorName ) == 'object' ) {
				this.cookieFactors.push( factor.factorName[0] ); // Some factors have multiple names
			} else {
				this.cookieFactors.push( factor.factorName );
			}
		} else if ( typeof( factor.factorName ) == 'object' ) {
			for ( var i=0; i<this.cookieFactors.length; i++ ) {
				if ( factor.factorName.inArray( this.cookieFactors[i] )) {
					this.cookieFactors.splice( i, 1 );
				}
			}
		} else {
			for ( var i=0; i<this.cookieFactors.length; i++ ) {
				if ( this.cookieFactors[i] == factor.factorName ) {
					this.cookieFactors.splice( i, 1 );
				}
			}
		}

		this.setCookie( this.factorCookie, this.cookieFactors.join());
	},

	saveAttribute: function( id, atName, atValue ) {
		for ( var i=0; i<this.cookieAttribs.length; i++ ) {
			if ( this.cookieAttribs[i].indexOf( id + '~' + atName ) != -1 ) {
				this.forgetAttribute( id, atName );
			}
		}

		this.cookieAttribs.push( id + '~' + atName + '~' + atValue );
		this.setCookie( this.attribsCookie, this.cookieAttribs.join());
	},

	forgetAttribute: function( id, atName ) {
		for ( var i=0; i<this.cookieAttribs.length; i++ ) {
			if ( this.cookieAttribs[i].indexOf( id + '~' + atName ) != -1 ) {
				this.cookieAttribs.splice( i, 1 );
			}
		}

		this.setCookie( this.attribsCookie, this.cookieAttribs.join());
	},

	closeInfo: function() {
		var fades = Array();
		$H( this.infoBoxes ).each( function(infoBox ) {
			fades.push( Effect.Fade( infoBox.value.infoID ));
		});

		new Effect.Parallel( fades, {duration: 0.20} );
	},

	showInfo: function( showInfoID ) {
		var fades = Array();
		$H( this.infoBoxes ).each( function( infoBox ) {
			if ( infoBox.value.infoID != showInfoID ) {
				fades.push( Effect.Fade( infoBox.value.infoID ));
			}
		});

		new Effect.Parallel(fades,{duration: 0.20, afterFinish: function(){Effect.Appear(showInfoID,{duration: 0.20});}});
		
	},

	enDisableCheck: function( trigger, inputEl, label ) {
		var trigger = $( trigger );

		if ( trigger.checked ) {
			$( inputEl ).setAttribute( 'disabled', 'disabled' );
			$( inputEl ).value='';
			$( label ).setAttribute( 'disabled', 'disabled' ); // ie
			$( label ).setAttribute( 'style', 'color: #999' ); // others
			this.saveAttribute( inputEl, 'disabled', 'disabled' );
			this.saveAttribute( label, 'disabled', 'disabled' );
			this.saveAttribute( label, 'style', 'color: #999' );
			this.saveAttribute( trigger.id, 'checked', 'checked' );
		} else {
			$( inputEl ).removeAttribute( 'disabled' );
			$( label ).removeAttribute( 'disabled' ); // ie
			$( label ).removeAttribute( 'style' ); // others
			this.forgetAttribute( inputEl, 'disabled' );
			this.forgetAttribute( label, 'disabled' );
			this.forgetAttribute( label, 'style' );
			this.forgetAttribute( trigger.id, 'checked' );
		}
	},

	enDisableKey: function( trigger, inputEl, label ) {
		var trigger = $( trigger );

		if ( trigger.value.length > 0 ) {
			$( inputEl ).setAttribute( 'disabled', 'disabled' );
			$( label ).setAttribute( 'disabled', 'disabled' ); // ie
			$( label ).setAttribute( 'style', 'color: #999' ); // others
		} else {
			$( inputEl ).removeAttribute( 'disabled' );
			$( label ).removeAttribute( 'disabled' ); // ie
			$( label ).removeAttribute( 'style' ); // others
		}
	}
});
